srcjs/shinyapp.js

```diff
@@ -714,6 +714,131 @@ var ShinyApp = function() {
     });
   });
 
+  addMessageHandler('shiny-insert-tab', function(message) {
+    var $tabsetPanel = $("#" + message.tabsetPanelId);
+    if ($tabsetPanel.length === 0) {
+      throw 'There is no tabsetPanel with id ' + message.tabsetPanelId;
+    };
+
+    // This is the JS equivalent of the builtItem() R function that is used
```

Comment:
>  I assume the answer to Winston's other question about `navbarPage`, `navbarMenu`, and `navlistPanel` would be a factor

It is and it isn't. So, in answer to @wch's question about whether this works for `navbarPage`, `navbarMenu`, and `navlistPanel`, the answer is no. I guess I assumed `tabsetPanel` was the only tabset-thingie that was worth it to implement this for (i.e. the one that people reasonably expect to be able to add and remove tabs). Maybe that's wrong. And this is where we get to @jcheng5's point. Accommodating these scenarios would involve changing my JS (since I only implemented the `tabsetPanel` case). This would again be mostly a translation of the R code.

So, if we want to accommodate `navbarPage`, `navbarMenu`, and `navlistPanel`, I think we really have to stop the code duplication and settle on a place/language to put this in. 
Like Winston, I think it makes more sense for this to be in the JS side, since that language is perfect for DOM mutations, whereas R isn't.

@wch , @jcheng5, should I go ahead with that? I think the whole refactoring + `navbarPage`, `navbarMenu`, and `navlistPanel` implementations + code review would likely take one more day (assuming all other feedback on the PR is already here).